Python/import.c

```diff
@@ -3963,6 +4211,193 @@ PyImport_ImportModuleLevelObject(PyObject *name, PyObject *globals,
+        Py_XDECREF(parent_module);
+        parent_module = PyImport_GetModule(parent);
+        if (parent_module == NULL) {
+            if (PyErr_Occurred()) {
+                goto done;
+            }
+
+            // Record the child to be added when the parent is imported.
+            PyObject *lazy_submodules;
+            if (PyDict_GetItemRef(lazy_modules, parent, &lazy_submodules) < 0) {
+                goto done;
+            }
+            if (lazy_submodules == NULL) {
+                lazy_submodules = PySet_New(NULL);
+                if (lazy_submodules == NULL) {
+                    goto done;
+                }
+                if (PyDict_SetItem(lazy_modules, parent, lazy_submodules) < 0) {
+                    Py_DECREF(lazy_submodules);
+                    goto done;
+                }
+            }
+            if (PySet_Add(lazy_submodules, child) < 0) {
+                Py_DECREF(lazy_submodules);
+                goto done;
+            }
+            Py_DECREF(lazy_submodules);
+        } else {
+            Py_XDECREF(parent_dict);
+            parent_dict = get_mod_dict(parent_module);
+            if (parent_dict == NULL) {
+                goto done;
+            }
+            if (PyDict_CheckExact(parent_dict) && !PyDict_Contains(parent_dict, child)) {
+                PyObject *lazy_module_attr = _PyLazyImport_New(builtins, parent, child);
+                if (lazy_module_attr == NULL) {
+                    goto done;
+                }
+                if (PyDict_SetItem(parent_dict, child, lazy_module_attr) < 0) {
+                    Py_DECREF(lazy_module_attr);
+                    goto done;
+                }
+                Py_DECREF(lazy_module_attr);
+            }
+        }
+
+        Py_DECREF(name);
+        name = parent;
+        parent = NULL;
+    }
+
+done:
+    Py_XDECREF(parent_dict);
+    Py_XDECREF(parent_module);
+    Py_XDECREF(child);
+    Py_XDECREF(parent);
+    Py_XDECREF(name);
+    return ret;
+}
+
+PyObject *
+_PyImport_LazyImportModuleLevelObject(PyThreadState *tstate,
+                                      PyObject *name, PyObject *builtins,
+                                      PyObject *globals, PyObject *locals,
+                                      PyObject *fromlist, int level)
+{
+    PyObject *abs_name = get_abs_name(tstate, name, globals, level);
+    if (abs_name == NULL) {
+        return NULL;
+    }
+
+    PyInterpreterState *interp = tstate->interp;
+    assert(_PyEval_GetFrame()->f_globals == _PyEval_GetFrame()->f_locals); // should only be called in global scope
+
+    // Check if the filter disables the lazy import
+    PyObject *filter = FT_ATOMIC_LOAD_PTR_RELAXED(LAZY_IMPORTS_FILTER(interp));
+    if (filter != NULL) {
+        PyObject *modname;
+        if (PyDict_GetItemRef(globals, &_Py_ID(__name__), &modname) < 0) {
+            return NULL;
+        } else if (modname == NULL) {
+            modname = Py_None;
+        }
```

Comment:
This should be two separate `if`s right? because you might have `modname == NULL` if the name doesn't exist.